### PR TITLE
Movement by time

### LINKED ISSUE #34: Collision detection - player & plattform
# User Story
As a player
I want to be able to land on a plattform when jumped higher than the plattform
for that the developer must implement a collision detection between the player and the plattform and handle the event.

# Description
- player sould be able to jump higher than the plattforms
- player sould land on the plattform after jumping higher than the plattform

# To do's:
- [x] add collision detection
- [x] implement handling of the event

### PR DESCRIPTION
**Pull request by JoscaWij**
Based on #46 
Fixed #34 collision detection

⇾ Refactoring movement code

Happy gaming 👩‍🚀🌑🎮

Deployment: https://spacey-movementbytime-snghh8em.herokuapp.com/game
Storybook: https://spacey-movementbytime-snghh8em.herokuapp.com/storybook/?path=/story/game-game--game-play

✅ player can jump on platform and stays there
✅ player can jump again when landing on platform
✅ player can only jump on platform when 90% of his body is higher than the platform

❌ player starts on a platform. I will fix this later with the placement of the platform and the start player.top value.

### CODE REVIEW

File: client/src/game/Game.js [hunk 1]
```diff
@@ -9,9 +14,34 @@ function resizeCanvas(canvas) {
...
 function handleKeyDown(event) {
-  movePlayer(event.code, player);
-  rotatePlayer(event.code, player, DIRECTIONS);
+  keyState[event.code] = true;
+  const direction = directionKeyCodes[event.code];
+  if (direction) {
+    rotatePlayer(player, direction);
+  }
+
+  if (keyState[DIRECTIONKEYS.UP] && player.isAbleToJump & !player.isJumping) {
```
Comment (j-toscani) [inline client/src/game/Game.js]: Is the single "&" a typo? As far as I found out, it is only used for 'bit wise comparison' (https://www.w3schools.com/js/js_bitwise.asp).  If you use it on purpose, can you elaborate, why you use it here?

File: client/src/game/GameLoop.js [hunk 2]
```diff
@@ -24,34 +30,56 @@ export const player = {
...
-  player.offsetY += physics.gravity;
-  player.top += player.offsetY;
+  const isPlayerOnFloor = player.top > floor - player.height;
 
-  if (player.top > floor - player.height) {
-    player.offsetY = 0;
+  if (isPlayerOnFloor) {
     player.top = floor - player.height;
-    player.jumping = false;
+    player.isAbleToJump = true;
```
Comment (j-toscani) [inline client/src/game/GameLoop.js:74]: Well done! Wanted to ask that because of your timeout but you already thought of that!

**Review by j-toscani — COMMENTED** [review]
lgtm! 

PS: I dont yet understate the platform states and the handleing of gravity. I will have a look at that later :)

File: client/src/game/Game.js [hunk 3]
```diff
@@ -1,6 +1,11 @@
 import React, { useEffect, useRef } from "react";
-import { gameLoop, player, DIRECTIONS } from "./GameLoop";
-import movePlayer from "./movePlayer";
+import {
+  gameLoop,
+  player,
+  keyState,
+  DIRECTIONKEYS,
```
Comment (lmachens) [inline client/src/game/Game.js]: Usually, you separate uppercase constants with an underline:
https://javascript.info/variables#uppercase-constants

```suggestion
  DIRECTION_KEYS,
```

File: client/src/game/Game.js [hunk 4]
```diff
@@ -9,9 +14,34 @@ function resizeCanvas(canvas) {
...
+const directionKeyCodes = {
+  ArrowRight: DIRECTIONS.RIGHT,
+  ArrowLeft: DIRECTIONS.LEFT,
+  Space: DIRECTIONS.JUMPING,
+};
+
 function handleKeyDown(event) {
-  movePlayer(event.code, player);
-  rotatePlayer(event.code, player, DIRECTIONS);
+  keyState[event.code] = true;
```
Comment (lmachens) [inline client/src/game/Game.js]: If this indicates which keys are pressed, I would rename it:
```suggestion
  activeKeys[event.code] = true;
```
or 
```suggestion
  keysState[event.code] = true;
```
or
```suggestion
  keyStates[event.code] = true;
```

File: client/src/game/Game.js (hunk 1 above)
Comment (lmachens) [reply client/src/game/Game.js]: Isn't `isAbleToJump` connected to `isJumping`? 
If so, it's enough to check for `isAbleToJump`.

File: client/src/game/Game.js [hunk 5]
```diff
@@ -9,9 +14,34 @@ function resizeCanvas(canvas) {
...
+    player.isJumping = true;
+    player.isAbleToJump = false;
+    setTimeout(() => {
+      player.isJumping = false;
+    }, 300);
+  }
+}
+
+function handleKeyUp(event) {
+  keyState[event.code] = false;
```
Comment (lmachens) [inline client/src/game/Game.js]: You remember all keys (like `n`, `m`). 
Instead, you could check if the direction exists (if it is a used key).

File: client/src/game/Game.js [hunk 6]
```diff
@@ -23,10 +53,12 @@ const Game = (props) => {
 
     gameLoop(canvas);
 
-    window.addEventListener("keydown", handleKeyDown);
+    window.addEventListener("keydown", handleKeyDown, true);
```
Comment (lmachens) [inline client/src/game/Game.js]: What is `true` used for?

File: client/src/game/GameLoop.js [hunk 7]
```diff
@@ -1,15 +1,21 @@
...
 function clearCanvas(canvas, context) {
   context.clearRect(0, 0, canvas.width, canvas.height);
 }
 
-const physics = {
-  friction: 0.7,
-  gravity: 1,
+export const GRAVITY = 7;
+
+export let keyState = {};
```
Comment (lmachens) [inline client/src/game/GameLoop.js]:
```suggestion
export const keyState = {};
```

File: client/src/game/Game.js [hunk 8]
```diff
@@ -9,9 +14,34 @@ function resizeCanvas(canvas) {
   canvas.height = 667;
 }
 
+const directionKeyCodes = {
+  ArrowRight: DIRECTIONS.RIGHT,
```
Comment (lmachens) [inline client/src/game/Game.js]:
```suggestion
  [DIRECTIONKEYS.RIGHT]: DIRECTIONS.RIGHT,
```

**Review by lmachens — COMMENTED** [review]
gj!

File: client/src/game/GameLoop.js [hunk 9]
```diff
@@ -24,34 +30,56 @@ export const player = {
...
+  if (isPlayerOnFloor) {
     player.top = floor - player.height;
-    player.jumping = false;
+    player.isAbleToJump = true;
     if (player.direction === DIRECTIONS.JUMPING) {
       player.direction = DIRECTIONS.FRONT;
     }
   }
+
+  isPlayerOnAnyPlatform(player, plattforms);
```
Comment (lmachens) [inline client/src/game/GameLoop.js]: `isPlayerOnAnyPlatform` sounds like it returns a boolean.

File: client/src/game/Game.js (hunk 1 above)
Comment (JoscaWij) [reply client/src/game/Game.js]: @j-toscani 
that definitely was a typo 🙈

File: client/src/game/Game.js (hunk 6 above)
Comment (JoscaWij) [reply client/src/game/Game.js]: https://stackoverflow.com/questions/5657292/why-is-false-used-after-this-simple-addeventlistener-function

The 3rd param is called `useCapture` and related to the children of the element. [useCapture = true, The handler is set on the capturing phase. Events will get to it before getting to its children.] My canvas element doesn't have children, so I don't need it ;)

File: client/src/game/Game.js (hunk 1 above)
Comment (JoscaWij) [reply client/src/game/Game.js]: @lmachens 
totally right. I kicked it out 😉

**Review by lmachens — APPROVED** [review]
👾

**Pull request merged**